From a9c38f35f4749a55794562477f8977d833060f88 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@notvalid.com>
Date: Wed, 14 Oct 2026 16:47:45 -0700
Subject: [PATCH] Update legacy Script fixtures from the final Doged reference

Retain final Tobias legacy flags/rules coverage from reference9690833b8b6709b7a7837bc397f0d04bf3b885d1, including current signature-hash imports, transaction IDs and Dogecoin chain-ID block versions. Correct two strict-DER rejection expectations after the independently documented64-byte legacy DER repair: malformed Schnorr-shaped encodings fail SIG_DER before CHECKSIG evaluation.

diff --git a/src/test/data/script_tests.json b/src/test/data/script_tests.json
index 28c2b83ce799677af74e8780135c884b85c4f13d..b63b7c1b3c83e607e5f2d58d6c1f41c41ca0c05c 100644
--- a/src/test/data/script_tests.json
+++ b/src/test/data/script_tests.json
@@ -3975,14 +3975,14 @@
     "0x41 0x0df4be7f5fe74b2855b92082720e889038e15d8d747334fa3f300ef4ab1db1eea56aa83d1d60809ff6703791736be87cfb6cbc5c4036aeed3b4ea4e6dab3509001",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
     "LEGACY_RULES,STRICTENC",
-    "EVAL_FALSE",
+    "SIG_DER",
     "CHECKSIG Schnorr w/ STRICTENC disallowed on legacy"
 ],
 [
     "0x41 0x8d37e95e36718d7fbc8ffd63b0b4ebb89dd5fb683510a95345869399f810a8724a00e5cbbb2190205ffabac601d80bf110a67013521b7a7c02b2a51e07d723eb01",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIGVERIFY 1",
     "LEGACY_RULES,STRICTENC",
-    "CHECKSIGVERIFY",
+    "SIG_DER",
     "CHECKSIGVERIFY Schnorr w/ STRICTENC disallowed on legacy"
 ],
 
diff --git a/test/functional/feature_legacy_script_flags.py b/test/functional/feature_legacy_script_flags.py
index 3c07db952cc0b1c08819a927e3e84ae1b9b6eacf..3e131928d576c1f95644b9614dcb881217012f9f 100644
--- a/test/functional/feature_legacy_script_flags.py
+++ b/test/functional/feature_legacy_script_flags.py
@@ -9,10 +9,12 @@ from test_framework.address import (
     SCRIPTSIG_OP_TRUE,
 )
 from test_framework.blocktools import (
+    VERSION_CHAIN_ID_BITS,
     create_block,
     create_coinbase,
     make_conform_to_ctor,
 )
+from test_framework.hash import hash160
 from test_framework.key import ECKey
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
 from test_framework.p2p import P2PDataStore
@@ -27,10 +29,8 @@ from test_framework.script import (
     OP_RETURN,
     OP_TRUE,
     CScript,
-    SignatureHash,
-    SignatureHashForkId,
-    hash160,
 )
+from test_framework.signature_hash import SignatureHash, SignatureHashForkId
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.txtools import pad_tx
 from test_framework.util import assert_raises_rpc_error
@@ -57,12 +57,12 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
             return CScript([OP_HASH160, hash160(CScript(opcodes)), OP_EQUAL])
 
         def make_block_with_txs(node, txs):
-            [tx.rehash() for tx in txs]
+            [tx.txid_hex for tx in txs]
             block = create_block(
                 int(node.getbestblockhash(), 16),
                 create_coinbase(node.getblockcount() + 1),
             )
-            block.nVersion = 4
+            block.nVersion = VERSION_CHAIN_ID_BITS | 4
             block.vtx += txs
             make_conform_to_ctor(block)
             block.hashMerkleRoot = block.calc_merkle_root()
@@ -90,6 +90,8 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         # - SCRIPT_ENABLE_SIGHASH_FORKID
         # - SCRIPT_ENABLE_SCHNORR_MULTISIG
         # - SCRIPT_ENFORCE_SIGCHECKS (not tested here)
+        # - All other flags except SCRIPT_VERIFY_P2SH
+        #   (which the only mandatory flag on Dogecoin)
         # Standard flags:
         # - SCRIPT_VERIFY_SIGPUSHONLY
 
@@ -104,19 +106,22 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         tx = CTransaction()
         tx.vin = [CTxIn(COutPoint(int(cointxid, 16), 0), SCRIPTSIG_OP_TRUE)]
         tx.vout = [
+            CTxOut(2000, p2sh([OP_CHECKSIG])),
+            CTxOut(2000, p2sh([OP_CHECKSIG])),
+            CTxOut(2000, p2sh([OP_CHECKMULTISIG])),
             CTxOut(2000, p2sh([OP_CHECKSIG])),
             CTxOut(2000, p2sh([OP_CHECKMULTISIG])),
+            CTxOut(2000, p2sh([OP_CHECKMULTISIG])),
             CTxOut(2000, p2sh([OP_IF, OP_ENDIF])),
             CTxOut(2000, CScript([OP_TRUE])),
-            CTxOut(coinvalue - 20000, p2sh([OP_TRUE])),
+            CTxOut(coinvalue - 30000, p2sh([OP_TRUE])),
         ]
-        tx.rehash()
-        txid = tx.hash
+        txid = tx.txid_hex
 
         # Fork both nodes, CLEANSTACK not enforced on Dogecoin
         tx_noncleanstack = CTransaction()
         tx_noncleanstack.vin = [
-            CTxIn(COutPoint(int(txid, 16), 4), CScript([b"jonny", CScript([OP_TRUE])]))
+            CTxIn(COutPoint(int(txid, 16), 8), CScript([b"jonny", CScript([OP_TRUE])]))
         ]
         pad_tx(tx_noncleanstack)
 
@@ -126,7 +131,7 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
             nonlegacy_node,
             nonlegacy_peer,
             [tx, tx_noncleanstack],
-            "blk-bad-inputs, parallel script check failed",
+            "mandatory-script-verify-flag-failed (Stack size must be exactly one after execution)",
         )
         check_mined_txs_success(nonlegacy_node, nonlegacy_peer, [tx])
 
@@ -149,10 +154,10 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         tx_sighash_legacy = CTransaction(tx_sighash)
         tx_sighash_legacy.vin[0].scriptSig = CScript([txsig, public_key, script])
 
-        # SIGHASH_FORKID not allowed on Dogecoin
+        # SIGHASH_FORKID doesn't work on Dogecoin, OP_CHECKSIG uses legacy sighash
         assert_raises_rpc_error(
             -26,
-            "mandatory-script-verify-flag-failed (Illegal use of SIGHASH_FORKID)",
+            "mandatory-script-verify-flag-failed (Script evaluated without error but finished with a false/empty top stack element)",
             node.sendrawtransaction,
             tx_sighash_bip143.serialize().hex(),
         )
@@ -160,7 +165,7 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
             node,
             peer,
             [tx_sighash_bip143],
-            "blk-bad-inputs, parallel script check failed",
+            "mandatory-script-verify-flag-failed (Script evaluated without error but finished with a false/empty top stack element)",
         )
 
         # Legacy sighash not allowed on XEC
@@ -174,7 +179,7 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
             nonlegacy_node,
             nonlegacy_peer,
             [tx_sighash_legacy],
-            "blk-bad-inputs, parallel script check failed",
+            "mandatory-script-verify-flag-failed (Signature must use SIGHASH_FORKID)",
         )
 
         # Legacy sighash ok on Dogecoin
@@ -185,24 +190,156 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         nonlegacy_node.sendrawtransaction(tx_sighash_bip143.serialize().hex())
         check_mined_txs_success(nonlegacy_node, nonlegacy_peer, [tx_sighash_bip143])
 
+        # Test SIGHASH_BUG (0x20)
+        script = CScript([OP_CHECKSIG])
+        tx_sighash_bug = CTransaction()
+        pad_tx(tx_sighash_bug)
+        tx_sighash_bug.vin = [CTxIn(COutPoint(int(txid, 16), 1))]
+
+        (sighash, _) = SignatureHash(script, tx_sighash_bug, 0, 0x21)
+        txsig = private_key.sign_ecdsa(sighash) + b"\x21"
+        tx_sighash_bug.vin[0].scriptSig = CScript([txsig, public_key, script])
+
+        # Not allowed in mempool
+        assert_raises_rpc_error(
+            -26,
+            "non-mandatory-script-verify-flag (Signature hash type missing or not understood)",
+            node.sendrawtransaction,
+            tx_sighash_bug.serialize().hex(),
+        )
+        # But: 0x20 is allowed in blocks
+        check_mined_txs_success(node, peer, [tx_sighash_bug])
+
+        # Non-legacy mode forces known sighash type
+        assert_raises_rpc_error(
+            -26,
+            "mandatory-script-verify-flag-failed (Signature hash type missing or not understood)",
+            nonlegacy_node.sendrawtransaction,
+            tx_sighash_bug.serialize().hex(),
+        )
+        nonlegacy_peer = check_mined_txs_fail(
+            nonlegacy_node,
+            nonlegacy_peer,
+            [tx_sighash_bug],
+            "mandatory-script-verify-flag-failed (Signature hash type missing or not understood)",
+        )
+
+        # SIGHASH_BUG for OP_CHECKMULTISIG
+        script = CScript([OP_CHECKMULTISIG])
+        tx_multi_sighash_bug = CTransaction()
+        pad_tx(tx_multi_sighash_bug)
+        tx_multi_sighash_bug.vin = [CTxIn(COutPoint(int(txid, 16), 2))]
+
+        # Not allowed in mempool
+        (sighash, _) = SignatureHash(script, tx_multi_sighash_bug, 0, 0x21)
+        txsig = private_key.sign_ecdsa(sighash) + b"\x21"
+        tx_multi_sighash_bug.vin[0].scriptSig = CScript([0, txsig, 1, public_key, 1, script])
+        assert_raises_rpc_error(
+            -26,
+            "non-mandatory-script-verify-flag (Signature hash type missing or not understood)",
+            node.sendrawtransaction,
+            tx_multi_sighash_bug.serialize().hex(),
+        )
+        # But: 0x20 allowed in blocks in OP_CHECKMULTISIG
+        check_mined_txs_success(node, peer, [tx_multi_sighash_bug])
+
+        # Non-legacy mode forces known sighash type
+        assert_raises_rpc_error(
+            -26,
+            "mandatory-script-verify-flag-failed (Signature hash type missing or not understood)",
+            nonlegacy_node.sendrawtransaction,
+            tx_multi_sighash_bug.serialize().hex(),
+        )
+        nonlegacy_peer = check_mined_txs_fail(
+            nonlegacy_node,
+            nonlegacy_peer,
+            [tx_multi_sighash_bug],
+            "mandatory-script-verify-flag-failed (Signature hash type missing or not understood)",
+        )
+
+        # Test borked SIGHASH_FORKID with legacy signatures
+        # On Dogecoin, signatures with SIGHASH_FORKID are valid if found in
+        # blocks and if legacy signatures are used. Bork = fork + bug
+        script = CScript([OP_CHECKSIG])
+        tx_sighash_bork = CTransaction()
+        pad_tx(tx_sighash_bork)
+        tx_sighash_bork.vin = [CTxIn(COutPoint(int(txid, 16), 3))]
+
+        (sighash, _) = SignatureHash(script, tx_sighash_bork, 0, 0x41)
+        txsig = private_key.sign_ecdsa(sighash) + b"\x41"
+        tx_sighash_bork.vin[0].scriptSig = CScript([txsig, public_key, script])
+
+        # SIGHASH_FORKID not allowed in mempool on Dogecoin
+        assert_raises_rpc_error(
+            -26,
+            "non-mandatory-script-verify-flag (Illegal use of SIGHASH_FORKID)",
+            node.sendrawtransaction,
+            tx_sighash_bork.serialize().hex(),
+        )
+        # But: allowed in blocks, as long as legacy signatures are used
+        check_mined_txs_success(node, peer, [tx_sighash_bork])
+
+        # Non-legacy doesn't support legacy signatures
+        assert_raises_rpc_error(
+            -26,
+            "mandatory-script-verify-flag-failed (Signature must be zero for failed CHECK(MULTI)SIG operation)",
+            nonlegacy_node.sendrawtransaction,
+            tx_sighash_bork.serialize().hex(),
+        )
+        nonlegacy_peer = check_mined_txs_fail(
+            nonlegacy_node,
+            nonlegacy_peer,
+            [tx_sighash_bork],
+            "mandatory-script-verify-flag-failed (Signature must be zero for failed CHECK(MULTI)SIG operation)",
+        )
+
+        # SIGHASH_FORKID for OP_CHECKMULTISIG with legacy signatures
+        script = CScript([OP_CHECKMULTISIG])
+        tx_multi_sighash_bork = CTransaction()
+        pad_tx(tx_multi_sighash_bork)
+        tx_multi_sighash_bork.vin = [CTxIn(COutPoint(int(txid, 16), 4))]
+
+        (sighash, _) = SignatureHash(script, tx_multi_sighash_bork, 0, 0x41)
+        txsig = private_key.sign_ecdsa(sighash) + b"\x41"
+        tx_multi_sighash_bork.vin[0].scriptSig = CScript([0, txsig, 1, public_key, 1, script])
+        assert_raises_rpc_error(
+            -26,
+            "non-mandatory-script-verify-flag (Illegal use of SIGHASH_FORKID)",
+            node.sendrawtransaction,
+            tx_multi_sighash_bork.serialize().hex(),
+        )
+        check_mined_txs_success(node, peer, [tx_multi_sighash_bork])
+
+        assert_raises_rpc_error(
+            -26,
+            "mandatory-script-verify-flag-failed (Signature must be zero for failed CHECK(MULTI)SIG operation)",
+            nonlegacy_node.sendrawtransaction,
+            tx_multi_sighash_bork.serialize().hex(),
+        )
+        nonlegacy_peer = check_mined_txs_fail(
+            nonlegacy_node,
+            nonlegacy_peer,
+            [tx_multi_sighash_bork],
+            "mandatory-script-verify-flag-failed (Signature must be zero for failed CHECK(MULTI)SIG operation)",
+        )
+
         # Nulldummy enforced on Dogecoin, on XEC it's a bitfield
         tx_nulldummy = CTransaction()
         tx_nulldummy.vin = [
             CTxIn(
-                COutPoint(int(txid, 16), 1),
-                CScript([20, b"sig", 1, public_key, 1, CScript([OP_CHECKMULTISIG])]),
+                COutPoint(int(txid, 16), 5),
+                CScript([20, 0, public_key, 1, CScript([OP_CHECKMULTISIG])]),
             )
         ]
         tx_nulldummy.vout = [CTxOut(0, CScript([OP_RETURN]))]
         assert_raises_rpc_error(
             -26,
-            "mandatory-script-verify-flag-failed (Dummy CHECKMULTISIG argument must be zero)",
+            "non-mandatory-script-verify-flag (Dummy CHECKMULTISIG argument must be zero)",
             node.sendrawtransaction,
             tx_nulldummy.serialize().hex(),
         )
-        peer = check_mined_txs_fail(
-            node, peer, [tx_nulldummy], "blk-bad-inputs, parallel script check failed"
-        )
+        check_mined_txs_success(node, peer, [tx_nulldummy])
+
         assert_raises_rpc_error(
             -26,
             "mandatory-script-verify-flag-failed (Bitfield's bit out of the expected range)",
@@ -213,21 +350,21 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
             nonlegacy_node,
             nonlegacy_peer,
             [tx_nulldummy],
-            "blk-bad-inputs, parallel script check failed",
+            "mandatory-script-verify-flag-failed (Bitfield's bit out of the expected range)",
         )
 
         # MINIMALIF enforced on Dogecoin, not on XEC
         tx_minimalif = CTransaction()
         tx_minimalif.vin = [
             CTxIn(
-                COutPoint(int(txid, 16), 2),
+                COutPoint(int(txid, 16), 6),
                 CScript([OP_TRUE, 1337, CScript([OP_IF, OP_ENDIF])]),
             )
         ]
         pad_tx(tx_minimalif)
         assert_raises_rpc_error(
             -26,
-            "mandatory-script-verify-flag-failed (OP_IF/NOTIF argument must be minimal)",
+            "non-mandatory-script-verify-flag (OP_IF/NOTIF argument must be minimal)",
             node.sendrawtransaction,
             tx_minimalif.serialize().hex(),
         )
@@ -239,7 +376,7 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
 
         # Non-push opcodes allowed in scriptSigs on Dogecoin in blocks
         tx_signonpush = CTransaction()
-        tx_signonpush.vin = [CTxIn(COutPoint(int(txid, 16), 3), CScript([OP_NOP]))]
+        tx_signonpush.vin = [CTxIn(COutPoint(int(txid, 16), 7), CScript([OP_NOP]))]
         pad_tx(tx_signonpush)
         check_mined_txs_success(node, peer, [tx_signonpush])
         # Disallowed on XEC
@@ -247,7 +384,7 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
             nonlegacy_node,
             nonlegacy_peer,
             [tx_signonpush],
-            "blk-bad-inputs, parallel script check failed",
+            "mandatory-script-verify-flag-failed (Only push operators allowed in signatures)",
         )
 
 
diff --git a/test/functional/feature_legacy_script_rules.py b/test/functional/feature_legacy_script_rules.py
index e4b022a59876954e40017596e06834af7bde5571..784b5f5871a38748979940d329efe0bce3a5e61e 100644
--- a/test/functional/feature_legacy_script_rules.py
+++ b/test/functional/feature_legacy_script_rules.py
@@ -9,7 +9,12 @@ from test_framework.address import (
     SCRIPT_UNSPENDABLE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import create_block, create_coinbase
+from test_framework.blocktools import (
+    VERSION_CHAIN_ID_BITS,
+    create_block,
+    create_coinbase,
+)
+from test_framework.hash import hash160
 from test_framework.key import ECKey
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
 from test_framework.p2p import P2PDataStore
@@ -34,10 +39,8 @@ from test_framework.script import (
     OP_SPLIT,
     OP_XOR,
     CScript,
-    SignatureHash,
-    SignatureHashForkId,
-    hash160,
 )
+from test_framework.signature_hash import SignatureHash, SignatureHashForkId
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.txtools import pad_tx
 from test_framework.util import assert_raises_rpc_error
@@ -64,12 +67,11 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
             return CScript([OP_HASH160, hash160(CScript(opcodes)), OP_EQUAL])
 
         def make_block_with_tx(node, tx):
-            tx.rehash()
             block = create_block(
                 int(node.getbestblockhash(), 16),
                 create_coinbase(node.getblockcount() + 1),
             )
-            block.nVersion = 4
+            block.nVersion = VERSION_CHAIN_ID_BITS | 4
             block.vtx += [tx]
             block.hashMerkleRoot = block.calc_merkle_root()
             block.solve()
@@ -147,7 +149,10 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
                 spend_tx.serialize().hex(),
             )
             peer = check_mined_tx_fails(
-                node, peer, spend_tx, "blk-bad-inputs, parallel script check failed"
+                node,
+                peer,
+                spend_tx,
+                "mandatory-script-verify-flag-failed (Attempted to use a disabled opcode)",
             )
             del spend_tx
 
@@ -167,7 +172,10 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
                 spend_if_tx.serialize().hex(),
             )
             peer = check_mined_tx_fails(
-                node, peer, spend_if_tx, "blk-bad-inputs, parallel script check failed"
+                node,
+                peer,
+                spend_if_tx,
+                "mandatory-script-verify-flag-failed (Attempted to use a disabled opcode)",
             )
             # Non-legacy nodes don't have these opcodes disabled
             nonlegacy_node.sendrawtransaction(spend_if_tx.serialize().hex())
@@ -193,7 +201,10 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
                 spend_tx.serialize().hex(),
             )
             peer = check_mined_tx_fails(
-                node, peer, spend_tx, "blk-bad-inputs, parallel script check failed"
+                node,
+                peer,
+                spend_tx,
+                "mandatory-script-verify-flag-failed (Opcode missing or not understood)",
             )
             del spend_tx
 
@@ -228,15 +239,18 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         txsig = private_key.sign_schnorr(sighash) + b"\x01"
         spend_tx.vin[0].scriptSig = CScript([txsig, public_key, script])
 
-        # Regtest has SCRIPT_VERIFY_NULLFAIL, which comes in handy here
+        # We don't get a "Signature must be zero" error as it's not mandatory
         assert_raises_rpc_error(
             -26,
-            "mandatory-script-verify-flag-failed (Signature must be zero for failed CHECK(MULTI)SIG operation)",
+            "mandatory-script-verify-flag-failed (Script evaluated without error but finished with a false/empty top stack element)",
             node.sendrawtransaction,
             spend_tx.serialize().hex(),
         )
         peer = check_mined_tx_fails(
-            node, peer, spend_tx, "blk-bad-inputs, parallel script check failed"
+            node,
+            peer,
+            spend_tx,
+            "mandatory-script-verify-flag-failed (Script evaluated without error but finished with a false/empty top stack element)",
         )
         # Schnorr sigs allowed on XEC with FORKID
         sighash = SignatureHashForkId(script, spend_tx, 0, 0x41, 2000)
@@ -256,12 +270,15 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
 
         assert_raises_rpc_error(
             -26,
-            "mandatory-script-verify-flag-failed (Signature must be zero for failed CHECK(MULTI)SIG operation)",
+            "mandatory-script-verify-flag-failed (Script failed an OP_CHECKSIGVERIFY operation)",
             node.sendrawtransaction,
             spend_tx.serialize().hex(),
         )
         peer = check_mined_tx_fails(
-            node, peer, spend_tx, "blk-bad-inputs, parallel script check failed"
+            node,
+            peer,
+            spend_tx,
+            "mandatory-script-verify-flag-failed (Script failed an OP_CHECKSIGVERIFY operation)",
         )
         sighash = SignatureHashForkId(script, spend_tx, 0, 0x41, 2000)
         txsig = private_key.sign_schnorr(sighash) + b"\x41"
